### PR TITLE
Issue 5944:  Improve createReaderGroup and getReaderGroup APIs

### LINKED ISSUE #5944: Ensure createReaderGroup API indicates if the reader group already exists.
**Describe the bug**
Ensure createReaderGroup API indicates if the ReaderGroup (RG) already exists. The user of the current API is not able to check if RG was created or an existing RG was used.

**To Reproduce**
1. Create RG with config1
2. Attempt creating RG with config2 with the same RG name as in Step 1.
There is no indication that the step 2 was not successful.

**Screenshots**
N/A

**Additional information**
The controller tracks the RG configuration and the _RGname stream creation. If the RG is already created then there is no need to attempt initializing the _RG stream. (This can be avoided).

### PR DESCRIPTION
**Pull request by shrids**
**Change log description**  
- Ensure `getReaderGroup` API throws a `ReaderGroupNotFoundException `if the ReaderGroup does not exist.
- Ensure `createReaderGroup` API indicates if the reader group already exists.

**Purpose of the change**  
Fixes #5944 

**What the code does**  
- Ensure `getReaderGroup` API throws a `ReaderGroupNotFoundException `if the ReaderGroup(RG) does not exist.
- Ensure `createReaderGroup` API indicates if the RG already exists. 

```
// Updated API signature to create a new RG.
  /**
     * Creates a new ReaderGroup.
     *
     * Readers will be able to join the group by calling
     * {@link io.pravega.client.EventStreamClientFactory#createReader(String, String, Serializer, ReaderConfig)}
     * . Once this is done they will start receiving events from the point defined in the config
     * passed here.
     * <p>
     * Note: This method is idempotent assuming called with the same name and config. This method
     * may block.
     * @param groupName The name of the group to be created.
     * @param config The configuration for the new ReaderGroup.
     * @return True if ReaderGroup is created and false if it already exists.
     * @throws ConfigMismatchException If the reader group already exists with a different configuration. Use 
     * {@link ReaderGroup#resetReaderGroup} to change the reader group configuration.
     * 
     */
    boolean createReaderGroup(String groupName, ReaderGroupConfig config) throws ConfigMismatchException;
```

**How to verify it**  
All the existing and the newly added tests should pass.

### CODE REVIEW

File: client/src/main/java/io/pravega/client/admin/impl/ReaderGroupManagerImpl.java [hunk 1]
```diff
@@ -95,11 +95,20 @@ public void createReaderGroup(String groupName, ReaderGroupConfig config) {
...
         }
         ReaderGroupConfig controllerConfig = getThrowingException(controller.createReaderGroup(scope, groupName, config));
-        @Cleanup
-        StateSynchronizer<ReaderGroupState> synchronizer = clientFactory.createStateSynchronizer(NameUtils.getStreamForReaderGroup(groupName),
-                                              new ReaderGroupStateUpdatesSerializer(), new ReaderGroupStateInitSerializer(), SynchronizerConfig.builder().build());
-        Map<SegmentWithRange, Long> segments = ReaderGroupImpl.getSegmentsForStreams(controller, controllerConfig);
-        synchronizer.initialize(new ReaderGroupState.ReaderGroupStateInit(controllerConfig, segments, getEndSegmentsForStreams(controllerConfig), false));
+        if (!controllerConfig.equals(config)) {
+            log.warn("ReaderGroup {} already exists with pre-existing configuration {}", groupName, controllerConfig);
+            return false;
```
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/admin/impl/ReaderGroupManagerImpl.java:100]: Should this throw?

File: client/src/main/java/io/pravega/client/admin/ReaderGroupManager.java [hunk 2]
```diff
@@ -71,8 +72,9 @@ public static ReaderGroupManager withScope(String scope, ClientConfig clientConf
      * may block.
      * @param groupName The name of the group to be created.
      * @param config The configuration for the new ReaderGroup.
+     * @return True if ReaderGroup is created with the provided config.
```
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/admin/ReaderGroupManager.java:75]: "is created" read to me as "was created" IE in response to this call. So I missed the meaning of this change the first time. We should either clarify the docs or change the behavior.

I would prefer the later, returning false vs true for something that is idempotent makes sense. But if it already exists and the passed config is different that isn't a case for idempotence. That seems like an exceptional case to me.

**Review by tkaitchuck — CHANGES_REQUESTED** [review]

File: client/src/main/java/io/pravega/client/admin/impl/ReaderGroupManagerImpl.java (hunk 1 above)
Comment (shrids) [reply client/src/main/java/io/pravega/client/admin/impl/ReaderGroupManagerImpl.java]: improved the code.

File: client/src/main/java/io/pravega/client/admin/ReaderGroupManager.java (hunk 2 above)
Comment (shrids) [reply client/src/main/java/io/pravega/client/admin/ReaderGroupManager.java]: I have improved the API signature based on this feedback.

File: client/src/main/java/io/pravega/client/admin/ReaderGroupManager.java [hunk 3]
```diff
@@ -88,6 +90,7 @@ public static ReaderGroupManager withScope(String scope, ClientConfig clientConf
      * 
      * @param groupName The name of the group
      * @return Reader group with the given name
+     * @throws ReaderGroupNotFoundException If the reader group does not exist.
      */
     ReaderGroup getReaderGroup(String groupName);
```
[the remark this answers is not in the document]
Comment (shrids) [reply client/src/main/java/io/pravega/client/admin/ReaderGroupManager.java]: ok

**Review by tkaitchuck — APPROVED** [review]

File: client/src/main/java/io/pravega/client/stream/ConfigMismatchException.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ */
+package io.pravega.client.stream;
+
+
+public class ConfigMismatchException extends IllegalArgumentException {
+    private static final long serialVersionUID = 1L;
+    public ConfigMismatchException(String readerGroupScopedName, ReaderGroupConfig currentConfig) {
+        super(String.format("Reader Group %s already exists with a different configuration %s.", readerGroupScopedName, currentConfig));
+    }
+
```
Comment (RaulGracia) [inline client/src/main/java/io/pravega/client/stream/ConfigMismatchException.java:24]: Nit: remove extra blank lines.

File: client/src/main/java/io/pravega/client/stream/ConfigMismatchException.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.pravega.client.stream;
+
+
```
Comment (RaulGracia) [inline client/src/main/java/io/pravega/client/stream/ConfigMismatchException.java:18]: Please, add comments describing this exception (what it is and when it is thrown).

File: client/src/main/java/io/pravega/client/stream/ConfigMismatchException.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+public class ConfigMismatchException extends IllegalArgumentException {
+    private static final long serialVersionUID = 1L;
+    public ConfigMismatchException(String readerGroupScopedName, ReaderGroupConfig currentConfig) {
```
Comment (RaulGracia) [inline client/src/main/java/io/pravega/client/stream/ConfigMismatchException.java:21]: Nit: missing blank line before constructor.

**Review by RaulGracia — CHANGES_REQUESTED** [review]
LGTM, just a couple of minor comments before merging.

File: client/src/main/java/io/pravega/client/stream/ConfigMismatchException.java (hunk 4 above)
Comment (shrids) [reply client/src/main/java/io/pravega/client/stream/ConfigMismatchException.java]: fixed it.

File: client/src/main/java/io/pravega/client/stream/ConfigMismatchException.java (hunk 5 above)
Comment (shrids) [reply client/src/main/java/io/pravega/client/stream/ConfigMismatchException.java]: done,improved other exception classes too.

File: client/src/main/java/io/pravega/client/stream/ConfigMismatchException.java (hunk 6 above)
Comment (shrids) [reply client/src/main/java/io/pravega/client/stream/ConfigMismatchException.java:25]: fixed it

**Review by RaulGracia — APPROVED** [review]

**Pull request merged**